pkg/auth/service_test.go

```diff
@@ -2605,3 +2605,135 @@ func TestAPIAuthService_AddCredentials(t *testing.T) {
+			if !errors.Is(err, tt.expectedErr) {
+				t.Fatalf("CreateUserExternalPrincipal: expected err: %v got: %v", tt.expectedErr, err)
+			}
+			if err != nil {
+				return
+			}
+		})
+	}
+}
+func TestAPIAuthService_ReusePrincipalAfterDelete(t *testing.T) {
+	mockClient, s := NewTestApiService(t, false)
+	userA := "user_a"
+	userB := "user_b"
+	principalId := "arn:aws:sts::123:assumed-role/MyRole/SessionName"
+	ctx := context.Background()
+
+	// create principal A for user1
+	mockClient.EXPECT().CreateUserExternalPrincipalWithResponse(gomock.Any(), userA, principalId).Return(&auth.CreateUserExternalPrincipalResponse{
+		HTTPResponse: &http.Response{
+			StatusCode: http.StatusCreated,
+		},
+	}, nil)
+	err := s.CreateUserExternalPrincipal(ctx, userA, principalId)
+	require.NoErrorf(t, err, "creating initial principal for user %s", userA)
+
+	// delete principal A for user1
+	mockClient.EXPECT().DeleteUserExternalPrincipalWithResponse(gomock.Any(), userA, principalId).Return(&auth.DeleteUserExternalPrincipalResponse{
+		HTTPResponse: &http.Response{
+			StatusCode: http.StatusNoContent,
+		},
+	}, nil)
+	err = s.DeleteUserExternalPrincipal(ctx, userA, principalId)
+	require.NoErrorf(t, err, "deleting principal for user %s", userA)
+
+	// re-use principal A again for user2
+	mockClient.EXPECT().CreateUserExternalPrincipalWithResponse(gomock.Any(), userB, principalId).Return(&auth.CreateUserExternalPrincipalResponse{
+		HTTPResponse: &http.Response{
+			StatusCode: http.StatusCreated,
+		},
+	}, nil)
+	err = s.CreateUserExternalPrincipal(ctx, userB, principalId)
+	require.NoErrorf(t, err, "re-using principal for user %s", userB)
+}
+
+func TestAPIAuthService_DeleteExternalPrincipalAttachedToUserDelete(t *testing.T) {
+	mockClient, s := NewTestApiService(t, false)
+	userId := "user"
+	principalId := "arn:aws:sts::123:assumed-role/MyRole/SessionName"
+	ctx := context.Background()
+
+	// create userA and principalA
+	mockClient.EXPECT().CreateUserExternalPrincipalWithResponse(gomock.Any(), userId, principalId).Return(&auth.CreateUserExternalPrincipalResponse{
+		HTTPResponse: &http.Response{
+			StatusCode: http.StatusCreated,
+		},
+	}, nil)
+	err := s.CreateUserExternalPrincipal(ctx, userId, principalId)
+	require.NoError(t, err)
+
+	// delete user A
+	mockClient.EXPECT().DeleteUserWithResponse(gomock.Any(), userId).Return(&auth.DeleteUserResponse{
+		HTTPResponse: &http.Response{
+			StatusCode: http.StatusNoContent,
+		},
+	}, nil)
+	err = s.DeleteUser(ctx, userId)
+	require.NoError(t, err)
+	// get principalA and expect error
+	mockClient.EXPECT().GetUserExternalPrincipalWithResponse(gomock.Any(), userId, gomock.Any()).Return(
+		&auth.GetUserExternalPrincipalResponse{
+			HTTPResponse: &http.Response{
+				StatusCode: http.StatusNotFound,
+			},
+		}, auth.ErrNotFound)
+
+	_, err = s.GetUserExternalPrincipal(ctx, userId, principalId)
+	require.Errorf(t, err, "principal should not exist if a user is deleted")
+}
```

Comment:
Nice!
I'm missing these tests:
1. ListUserExternalPrincipals and GetExternalPrincipal- unless you think it's pointless because it will be a "mocking test"
2. Delete external principal for deleted user should work

Reply:
I'm also missing some RBAC tests

Reply:
> ListUserExternalPrincipals and GetExternalPrincipal- unless you think it's pointless because it will be a "mocking test"`

I think it's pointless 👍 

> Delete external principal for deleted user should work

Not it's not 👀  
When calling it will fail if the user doesn't exist and that's a good thing! (same as our other flows such as `DetachPolicyFromUser`)